File: android/src/main/kotlin/africa/ejara/trustdart/WalletHandler.kt
package africa.ejara.trustdart

import BTC
import ETH
import NEAR
import SOL
import TRX
import XTZ
import XLM
import BNB
import DOGE
import MATIC
import ADA

import africa.ejara.trustdart.utils.WalletError
import africa.ejara.trustdart.utils.WalletValidateResponse
import wallet.core.jni.CoinType
import wallet.core.jni.HDWallet

class WalletHandler {

    companion object {
        val coins = mapOf(
            "BTC"   to BTC(),
            "ETH"   to ETH(),
            "XTZ"   to XTZ(),
            "TRX"   to TRX(),
            "SOL"   to SOL(),
            "NEAR"  to NEAR(),
            "XLM"   to XLM(),
            "BNB"   to BNB(),
            "BSC"   to ETH(),
            "DOGE"  to DOGE(),
            "MATIC" to ETH(),
            "ADA"   to ADA(),
        )
    }
    
    fun getCoin(coin: String?): Coin {
        return coins[coin]!!
    }

    fun generateMnemonic(strength: Int, passphrase: String): String? {
        return HDWallet(strength, passphrase).mnemonic()
    }

    fun checkMnemonic(mnemonic: String, passphrase: String): HDWallet {
        return HDWallet(mnemonic, passphrase)
    }

    fun <T> validate(walletError: WalletError, data: Array<T>): WalletValidateResponse {
        var isValid = true
        for (_data in data) {
            if (_data == null) {
                isValid = false
            }
        }
        return WalletValidateResponse(isValid, walletError)
    }

}

File: ios/Classes/WalletHandler.swift
//
//  WalletHandler.swift
//  Pods-Runner
//
//  Created by Jay on 3/29/22.
//

import WalletCore
import Flutter

class WalletHandler {
    static let coins: [String: Coin] = [
        "BTC"   : BTC(),
        "ETH"   : ETH(),
        "XTZ"   : XTZ(),
        "TRX"   : TRX(),
        "SOL"   : SOL(),
        "NEAR"  : NEAR(),
        "XLM"   : XLM(),
    ]
    
    func getCoin(_ coin: String) -> Coin {
        return WalletHandler.coins[coin]!
    }
    
    func generateMnemonic(strength: Int32, passphrase: String) -> String? {
        return HDWallet(strength: strength, passphrase: passphrase)?.mnemonic
    }
    
    func checkMnemonic(mnemonic: String?, passphrase: String?) -> HDWallet?{
        return HDWallet(mnemonic: mnemonic!, passphrase: passphrase!)
    }
    
    static func validate<T>(walletError: WalletError, _ data: T?...) -> (Bool, WalletError) {
        var isValid = true
        data.forEach { _data in
            if _data == nil {
                isValid = false
            }
        }
        return (isValid, walletError.self)
    }
}


